src/utils.jl

```diff
@@ -313,50 +202,104 @@ function split_dataset(data::Matrix; train_ratio::Float64 = 0.9, randomize::Bool
   return train_data,test_data
 end
 
+"""
+    _ischoi(M::LPDO)
 
+Check whether a given LPDO{MPO}  
+"""
+_ischoi(M::LPDO{MPO}) = (length(inds(M.X[1],"Site")) == 2 && hastags(M.X[1],"Purifier"))
```

Comment:
I think we need to be clearer about what we define to be Choi. I would say it would be an `LPDO{MPO}` `L` where the tensors of `L.X` have site indices that comes in pairs with tags `"Input"` and `"Output"` (so essentially `haschoitags(L.X)`). I don't think we should reference "Purifier" here.

Then, I would define `ischoi(::MPO)` in a similar where, but where the site indices also come prime and unprimed pairs.

We can make the definitions better defined later if these definitions work right now.

Reply:
Right, but also an MPO with 4 site indices (like the Choi matrix generate by `runcircuit`).

```julia
ischoi(M::LPDO{MPO}) = (length(inds(M.X[1],"Site")) == 2 && haschoitags(M))
ischoi(M::MPO) = (length(inds(M[1],"Site")) == 4 && haschoitags(M))
```
So now we check that the object has the Choi tags, but also that the index structure is correct. That is, `M` has 2 site indices if it is a `LPDO{MPO}` and 4 site indices if it is an `MPO`.

Reply:
Sure. In theory we should check all of the tensors but again that is fine for now.